### PR TITLE
`ultralytics 8.3.216` Faster `Annotator` mask plotting

### PR DESCRIPTION
**Pull request by glenn-jocher**
Refactor mask handling and improve tensor operations in plotting.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improves segmentation mask rendering and stability, with a small version bump to 8.3.216. 🖼️✨

### 📊 Key Changes
- Bumped version: 8.3.215 → 8.3.216.
- Fixed tensor creation in `process_mask` by using `torch.tensor(..., device=...)` for correct device placement.
- Enhanced `plotting.Plotter.masks`:
  - Clearer assertion message.
  - Early return when no masks to avoid unnecessary ops.
  - Proper device alignment between `im_gpu` and `masks`.
  - For non-retina masks, now uses `ops.scale_masks` to remove padding and upsample correctly.
  - Converts the source image from BGR to RGB tensor when needed.
  - Streamlined blending pipeline (fewer conversions, more efficient tensor ops).

### 🎯 Purpose & Impact
- More accurate and consistent mask overlays, especially when `retina_masks=False` (padding is correctly removed). ✅
- Fewer runtime/device mismatch issues on GPU thanks to safer tensor device handling. 🛡️
- Slight performance gains from reduced conversions and intermediate allocations. ⚡
- Better user experience in visualizations across segmentation workflows in Ultralytics models. 🎨

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response — an Ultralytics engineer will review and assist you shortly. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional notes for this PR 🎨
- Please verify behavior with both `retina_masks=True` and `retina_masks=False`, across CPU/GPU, to ensure consistent mask scaling, color blending, and channel handling.
- Consider adding unit tests for the updated paths in `ultralytics/utils/plotting.py`, including cases with no masks, different image sizes, and device alignment between `im_gpu` and `masks`.
- Double-check that early returns and normalization do not introduce regressions in downstream visualization flows.
- If public behavior changes (e.g., mask rendering differences), add a short note in docs/changelog for clarity.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by glenn-jocher** [comment]
### Main
 
```
yolo predict imgsz=320 model=yolo11n-seg.pt save
Speed: 0.7ms preprocess, 12.8ms inference, 0.8ms postprocess per image at shape (1, 3, 192, 320)
```
image

### This PR
 
```
yolo predict imgsz=320 model=yolo11n-seg.pt save
Speed: 0.5ms preprocess, 14.9ms inference, 0.8ms postprocess per image at shape (1, 3, 192, 320)
```
image

**Comment by glenn-jocher** [comment]
Hmm, padding bug, fixing.

EDIT: Fixed!

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `optimize-mask-plotting` at 2025-10-16 12:38:39 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18561511664): `ci.yml` run 13319
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18561511873): `docker.yml` run 1306

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `optimize-mask-plotting` at 2025-10-16 13:15:04 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18562515757): `ci.yml` run 13322
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18562515963): `docker.yml` run 1307

**Comment by UltralyticsAssistant** [comment]
Fantastic work, @glenn-jocher — thrilled to see this merged! 

“Trifles make perfection, and perfection is no trifle.” — Michelangelo. Your refinement of mask rendering, device alignment, and the streamlined blending pipeline in [PR #22419](https://github.com/ultralytics/ultralytics/pull/22419) elevates the clarity and consistency of segmentation outputs while improving stability and performance. The bump to 8.3.216 brings tangible UX gains, especially with accurate non-retina mask handling.

Huge thanks for the craftsmanship and attention to detail—this makes a real difference for our community.
